imxrt-hal/src/can/embedded_hal.rs

```diff
@@ -0,0 +1,105 @@
+//! `embedded_hal` trait impls.
+
+use super::{Data, Frame, NoDataError, CAN};
+use crate::iomuxc::consts::Unsigned;
+
+use embedded_hal::can;
+pub use embedded_hal::can::{ExtendedId, Id, StandardId};
+
+impl<M> can::Can for CAN<M>
+where
+    M: Unsigned,
+{
+    type Frame = Frame;
+
+    type Error = NoDataError;
+
+    fn transmit(&mut self, frame: &Self::Frame) -> nb::Result<Option<Self::Frame>, Self::Error> {
+        match self.transmit(frame) {
+            Ok(_status) => Ok(Some(frame.clone())),
+            Err(nb::Error::WouldBlock) => Err(nb::Error::WouldBlock),
+            Err(nb::Error::Other(e)) => match e {},
+        }
+    }
+
+    fn receive(&mut self) -> nb::Result<Self::Frame, Self::Error> {
+        match self.read_mailboxes() {
+            Some(d) => Ok(d.frame),
+            None => Err(nb::Error::Other(NoDataError { _priv: () })),
+        }
+    }
+}
+
+impl can::Error for NoDataError {
+    fn kind(&self) -> can::ErrorKind {
+        can::ErrorKind::Overrun
+    }
+}
```

Comment:
Couple of questions / thoughts about our error handling:

- Can we not detect any other errors in this driver? `can::ErrorKind` has a variety of other errors that could help others with debugging.
- Is `Overrun` the proper kind of error for a "no data error?" My understanding of _overrun_ is "the buffer is full, and you tried to put more data in." This implies that there _is_ data available. If that's the right interpretation, maybe `ErrorKind::Other` would be more appropriate?
- We're using `NoDataError` as the error returns for both CAN transmit and receives (above). Since we need a single type to support multiple operations, should we consider a more generic `Error` enum, possibly non-exhaustive? This would let us incrementally add support for new errors without breaking users.

    ```rust
    /// FlexCAN-specific errors.
    #[non_exhaustive]
    pub enum Error {
      /// Receive buffer is empty, and there's no data.
      NoData,
    }

    impl can::Error for Error { /* ... */ }
    ```

Reply:
> Couple of questions / thoughts about our error handling:
> 
> * Can we not detect any other errors in this driver? `can::ErrorKind` has a variety of other errors that could help others with debugging.

Yes, that would be helpful - I will need to dig in a bit to the implementation to see where the other `ErrorKind`s may fit. 

> * ~Is `Overrun` the proper kind of error for a "no data error?" My understanding of _overrun_ is "the buffer is full, and you tried to put more data in." This implies that there _is_ data available. If that's the right interpretation, maybe `ErrorKind::Other` would be more appropriate?~ My understanding and interpretation may be wrong; `embedded_hal::can::ErrorKind::Overrun` is specifically for the _receive_ buffer. I suppose I'm not sure if `Overrun` is intended to tell the user "I received so much data from another device that I overran my buffer," or if it tells the user "there's no data in the receive buffer."

Yes, that is my interpretation too and I agree the naming is misleading. Updated to:

`can/mod.rs`

```rust
`/// Error from the FlexCan peripheral.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
/// Error indicating that no received data is 
/// available in the mailboxes
NoRxData,
/// Error indicating that no transmit mailboxes
/// are available
NoTxMailbox, 
/// A wrapper around the [`embedded_hal::ErrorKind`]
/// enum
EmbeddedHal(embedded_hal::ErrorKind),
}
```
```rust
#[inline(always)]
pub fn transmit(&mut self, frame: &Frame) -> nb::Result<(), Error> {
    for i in self.mailbox_offset()..self.get_max_mailbox() {
        if let FlexCanMailboxCSCode::TxInactive =
            FlexCanMailboxCSCode::from_code_reg(self.read_mailbox_code(i))
        {
            self.write_tx_mailbox(i, frame);
            return Ok(());
        }
    }
    Err(nb::Error::Other(Error::NoTxMailbox))
}
```

`can/embedded_hal.rs`

```rust
impl<M> can::Can for CAN<M>
where
    M: Unsigned,
{
    type Frame = Frame;

    type Error = Error;

    fn transmit(&mut self, frame: &Self::Frame) -> nb::Result<Option<Self::Frame>, Self::Error> {
        match self.transmit(frame) {
            Ok(_status) => Ok(Some(frame.clone())),
            Err(nb::Error::WouldBlock) => Err(nb::Error::WouldBlock),
            Err(nb::Error::Other(e)) => Err(nb::Error::Other(e)),
        }
    }

    fn receive(&mut self) -> nb::Result<Self::Frame, Self::Error> {
        match self.read_mailboxes() {
            Some(d) => Ok(d.frame),
            None => Err(nb::Error::Other(Error::NoRxData)),
        }
    }
}

impl can::Error for Error {
    fn kind(&self) -> can::ErrorKind {
        match self {
            Self::NoRxData => can::ErrorKind::Other,
            Self::NoTxMailbox => can::ErrorKind::Other,
            Self::EmbeddedHal(e) => e.kind()
        }
    }
}
```

> * We're using `NoDataError` as the error returns for both CAN transmit and receives (above). Since we need a single type to support multiple operations, should we consider a more generic `Error` enum, possibly non-exhaustive? This would let us incrementally add support for new errors without breaking users.
>   ```rust
>   /// FlexCAN-specific errors.
>   #[non_exhaustive]
>   pub enum Error {
>     /// Receive buffer is empty, and there's no data.
>     NoData,
>   }
>   
>   impl can::Error for Error { /* ... */ }
>   ```